src/lib/Libattr/attr_func.c

```diff
@@ -1294,45 +1287,77 @@ strtok_quoted(char *source, const char *delimiters)
 	if ((pc == NULL) || (*pc == '\0'))
 		return (NULL);
 
-	deli_len = (int)strlen(delimiters);
 	for (stok=pc; *pc != 0; pc++) {
 
-		for (i=0; i < deli_len; i++) {
-			if (*pc == delimiters[i]) {
-				break;
-			}
-		}
-		if (i < deli_len) { /* found a match */
+		/* must not match <ESC_CHAR><delim> or <ESC_CHAR><ESC_CHAR><delim>
+		 * the latter means <ESC_CHAR> is the one escaped not <delim>
+		 */
+		if ((*pc == delimiter) &&
+			(((pc-1) < stok) || (*(pc-1) != ESC_CHAR) ||
+			 ((pc-2) < stok) || (*(pc-2) == ESC_CHAR))) {
 			*pc = '\0';
 			pc++;
-			prune_quotes(stok);
+			prune_esc_backslash(stok);
 			return (stok);
 		}
 
-		/* check for a quoted value and advance */
-		/* pointer up to the closing quote. */
-
-		/* if a delimiter apears inside the */
-		/* quote, ex. "apple,bee", this will */
-		/* advance past the delimiter until */
-		/* until the matching quote is found. */
+		/* check for a quoted value and advance
+		 * pointer up to the closing quote. If a delimiter appears
+		 * inside the quote, ex. "apple,bee", this will
+		 * advance past the delimiter until the matching quote is
+		 * found followed by a delimiter * (e.g. ",) or
+		 * the matching quote followed by ending null character.
+		 */
 		if ((*pc == '\'') || (*pc == '"')) {
+
+			/* if immediately following the quote, try to match
+			 * one of:
+			 * 	'<null>
+			 *	'<delimiter>
+			 * 	"<null>
+			 *	"<delimiter>
+			 */
+			if ((*(pc+1) == '\0') || (*(pc+1) == delimiter)) {
+				pc++;
+				if (*pc != '\0') {
+					*pc = '\0';
+					pc++;
+				}
+				prune_esc_backslash(stok);
+				return (stok);
```

Comment:
The comment in the function that said "keep a quoted string together and not break it by a delimiter" is not right. The non-escaped delimiter takes precedence, and should stop there and return the token. This catches not balanced tokens. Also, if we continue to move past a non-escaped delimiter, it would end up eating the rest of the values if no ending quote is found. I'll update the comment.